### PR TITLE
Fix required field validator bug

### PR DESCRIPTION
**Pull request by dimagorelik**
Fix null pointer exception when field does not exist in entity change

### CODE REVIEW

File: main/src/main/java/com/kenshoo/pl/entity/internal/validators/RequiredFieldValidationAdapter.java [hunk 1]
```diff
@@ -36,9 +36,13 @@ public SupportedChangeOperation getSupportedChangeOperation() {
...
     public ValidationError validate(EntityChange<E> entityChange, CurrentEntityState currentState) {
-        if (entityChange.get(validator.requiredField()) == null && validator.requireWhen().test(currentState)) {
+        if (!isFieldSpecified(entityChange) && validator.requireWhen().test(currentState)) {
             return new ValidationError(validator.getErrorCode(), validator.requiredField(), ImmutableMap.of("field", validator.requiredField().toString()));
         }
         return null;
     }
+
+    private boolean isFieldSpecified(EntityChange<E> entityChange) {
+        return entityChange.isFieldChanged(validator.requiredField()) && entityChange.get(validator.requiredField()) != null;
```
Comment (effiban) [inline main/src/main/java/com/kenshoo/pl/entity/internal/validators/RequiredFieldValidationAdapter.java:46]: @dimagorelik - you can simplify this using `safeGet()`:
```
return entityChange.safeGet(validator.requiredField()).isPresent();
```

Comment (dimagorelik) [reply main/src/main/java/com/kenshoo/pl/entity/internal/validators/RequiredFieldValidationAdapter.java]: @effiban 
fixed

**Comment by dimagorelik** [comment]
retest this please

File: main/src/test/java/com/kenshoo/pl/entity/TriptionalTest.java [hunk 2]
```diff
@@ -114,6 +114,17 @@ public void absent_ShouldNotBeNull() {
...
 
+    @Test
+    public void nullInstance_ShouldBeNullOrAbsent() {
+        assertThat(Triptional.nullInstance().isNullOrAbsent(), is(true));
+    }
+
+    @Test
+    public void absent_ShouldBeNullOrAbsent() {
+        assertThat(Triptional.absent().isNullOrAbsent(), is(true));
+    }
```
Comment (effiban) [inline main/src/test/java/com/kenshoo/pl/entity/TriptionalTest.java:125]: what about testing `present`?

File: main/src/test/java/com/kenshoo/pl/entity/spi/helpers/EntityChangeCompositeValidatorTest.java [hunk 3]
```diff
@@ -155,7 +155,7 @@ public void registerEntityChangeValidatorTest() {
     public void registerRequiredFieldValidatorTest() {
         when(requiredFieldValidator.requiredField()).thenReturn(TestEntity.FIELD_1);
         when(requiredFieldValidator.requireWhen()).thenReturn(when-> true);
-        when(entityChange.get(TestEntity.FIELD_1)).thenReturn(null);
+        when(entityChange.safeGet(TestEntity.FIELD_1)).thenReturn(Triptional.of(null));
```
Comment (effiban) [inline main/src/test/java/com/kenshoo/pl/entity/spi/helpers/EntityChangeCompositeValidatorTest.java:158]: Here you can use `Triptional.nullInstance()`

**Review by effiban — APPROVED** [review]

File: main/src/main/java/com/kenshoo/pl/entity/internal/validators/RequiredFieldValidationAdapter.java [hunk 4]
```diff
@@ -36,9 +31,13 @@ public SupportedChangeOperation getSupportedChangeOperation() {
...
     public ValidationError validate(EntityChange<E> entityChange, CurrentEntityState currentState) {
-        if (entityChange.get(validator.requiredField()) == null && validator.requireWhen().test(currentState)) {
+        if (isFieldNotSpecified(entityChange) && validator.requireWhen().test(currentState)) {
             return new ValidationError(validator.getErrorCode(), validator.requiredField(), ImmutableMap.of("field", validator.requiredField().toString()));
         }
         return null;
     }
+
+    private boolean isFieldNotSpecified(EntityChange<E> entityChange) {
+        return  entityChange.safeGet(validator.requiredField()).isNullOrAbsent();
```
Comment (galkoren) [inline main/src/main/java/com/kenshoo/pl/entity/internal/validators/RequiredFieldValidationAdapter.java:41]: But `null` may be considered a valid value in the DB.
Suppose the `cpcBid=5` in DB and now I want to set it to `null`. It should be legal. And a validator should be able to cast logic to null values as well.

Comment (effiban) [reply main/src/main/java/com/kenshoo/pl/entity/internal/validators/RequiredFieldValidationAdapter.java:41]: I thought about that, but I assumed that required fields can't be `null`. @dimagorelik ?

Comment (dimagorelik) [reply main/src/main/java/com/kenshoo/pl/entity/internal/validators/RequiredFieldValidationAdapter.java:41]: @effiban @galkoren  
I didn't change the business logic (backward compatibility reason). The Null was considered as no value.

**Pull request merged**